src/inmanta/ast/__init__.py

```diff
@@ -251,9 +251,21 @@ def __init__(self, namespace: "Namespace", type: LocatableString) -> None:
         self.type = type
 
     def resolve(self) -> Optional[AnchorTarget]:
-        t = self.namespace.get_type(self.type)
-        location = t.get_location()
-        docstring = t.comment if isinstance(t, WithComment) else None
+        tp = None
+        try:
+            tp = self.namespace.get_type(self.type)
+        except TypeNotFoundException as e:
```

Comment:
I don't think this logic is entirely correct, I need to think on a better approach. The issue with this one is that it returns the first name it finds that matches, regardless of the type. Suppose you have the modules below. Then it would result in `b::C` rather than `c::C` (assuming that `b` would come before `c` in `visible_namespaces`, which I think is the case). The logic in `Constructor` would disregard `b::C` because it is not compatible with the type declared in the relation definition.

`a`:
```
entity A: end
A.c [1] -- c::C

A(c=C())
```

`b`:
```
entity B: end
entity C: end
```

`c`:
```
entity C: end
```

Reply:
indeed, I didn't think of that :( I'll try to find something better to

Reply:
How about simply move the anchor registration to the normalization call? Anchors are never fetched before normalization as far as I can see. This is not part of any docstring, so we would have to document it in e.g. the `get_anchors` method or something. @wouterdb CC.

```diff
diff --git a/src/inmanta/ast/statements/generator.py b/src/inmanta/ast/statements/generator.py
index a585269b..df03f3cc 100644
--- a/src/inmanta/ast/statements/generator.py
+++ b/src/inmanta/ast/statements/generator.py
@@ -820,7 +820,6 @@ class Constructor(ExpressionStatement):
         self.__wrapped_kwarg_attributes: List[WrappedKwargs] = wrapped_kwargs
         self.location = location
         self.namespace = namespace
-        self.anchors.append(TypeReferenceAnchor(namespace, class_type))
         for a in attributes:
             self.add_attribute(a[0], a[1])
         self.type: Optional["Entity"] = None
@@ -863,6 +862,7 @@ class Constructor(ExpressionStatement):

     def normalize(self, *, lhs_attribute: Optional[AttributeAssignmentLHS] = None) -> None:
         self.type = self._resolve_type(lhs_attribute)
+        self.anchors.append(TypeReferenceAnchor(self.namespace, self.type))

         inindex: abc.MutableSet[str] = set()
```

Reply:
My patch is not correct, but you get the point. It should be something more like `...(self.type.namespace, self.class_type)` I think.

Reply:
I agree that:
1. this is making a guess, which (up to this point), we never do in the compiler
2. I would not duplicate the resolution logic
3. I agree that we should make sure normalization is called before we get the anchors out and use the existing normalization code to set the anchors